pedsnetdcc/id_mapping_transform.py

```diff
@@ -0,0 +1,239 @@
+        # in a single transaction to prevent race conditions.
+        update_last_id_stmts = StatementList()
+        update_last_id_stmts.append(Statement(
+            lock_last_id_sql.format(**results),
+            lock_last_id_msg.format(**results)))
+        update_last_id_stmts.append(Statement(
+            update_last_id_sql.format(**results),
+            update_last_id_msg.format(**results)))
+
+        # Execute last id table update statements and ensure they didn't error
+        # and the second one returned results.
+        update_last_id_stmts.serial_execute(conn_str, transaction=True)
+
+        for stmt in update_last_id_stmts:
+            check_stmt_err(stmt, 'ID mapping pre-transform')
+        check_stmt_data(update_last_id_stmts[1], 'ID mapping pre-transform')
+
+        # Get the old and new last IDs from the second update statement.
+        results['old_last_id'] = update_last_id_stmts[1].data[0][0]
+        results['new_last_id'] = update_last_id_stmts[1].data[0][1]
+        logger.info({'msg': 'last ID tracking table updated',
+                     'table': table_name,
+                     'old_last})d': results['old_last_id'],
+                     'new_last_id': results['new_last_id']})
+
+        # Build statement to generate new ID maps.
+        insert_new_maps_stmt = Statement(
+            insert_new_maps_sql.format(**results),
+            insert_new_maps_msg.format(**results))
+
+        # Execute new map generation statement and ensure it didn't error.
+        insert_new_maps_stmt.execute(conn_str)
+        check_stmt_err(insert_new_maps_stmt, 'id mapping pre-transform')
+        logging.info({'msg': 'generated new ID mappings', 'table': table_name,
+                      'count': insert_new_maps_stmt.rowcount,
+                      'elapsed': secs_since(starttime)})
+
+    @classmethod
+    def modify_select(cls, metadata, table_name, select=None, join=None):
+        """Alter foreign key columns to get mapped DCC IDs.
+
+        The primary key and each foreign key which points at a data table
+        (i.e., not a vocabulary table) is altered to instead get the dcc_id's
+        from the mapping table, which requires a join to be added.
+        """
+
+        # Raise error if attempted on a vocabulary table.
+        if table_name in VOCAB_TABLES:
+            raise ValueError('ID mapping should not be performed on vocabulary'
+                             ' tables like {0}'.format(table_name))
+
+        # Get table object.
+        table = metadata.tables[table_name]
+
+        # Raise error if attempted on a multi-column primary key table.
+        if len(list(table.primary_key.columns.keys())) > 1:
+            raise ValueError('cannot map IDs for multi-column primary key'
+                             ' on table {0}'.format(table_name))
+
+        # Get primary key name and mapping table name, defined by convention.
+        pkey_name = list(table.primary_key.columns.keys())[0]
```

Comment:
What are we doing about `fact_relationship`, which has no PK, and `death`, which has `person_id` as its PK?  Aside: it looks like `death` doesn't have an FK on `person_id` into `person`, which it should.